### PR TITLE
storage: switch reclocking data format to rows

### PR DESCRIPTION
**Pull request by petrosagg**
This PR is the minimum surgery required in the current code that preserves all the reclocking logic as-is but changes the remap shard to be identically typed to the one of storage collections.

Reclock bindings for kafka sources are represented with two columns. An int32 for the partition id and a uin64 for the offset.

Reclock bindings for postgres sources are represented with just one column for the offset.

### CODE REVIEW

**Review by aljoscha — APPROVED** [review]
Yep, this is very neat! With `integrate()`/`differentiate()`. We could probably change the internal logic to also not store the offset in the diff field anymore but I also agree that this diff (pun intended!) is easier to review/be reasonably sure that it is correct.

Still scary, though, messing around with reclocking this shortly before the release cut! :sweat_smile:

File: src/storage/src/source/reclock.rs [hunk 1]
```diff
@@ -47,6 +47,91 @@ pub struct ReclockFollower {
...
+fn unpack_binding(data: SourceData) -> (PartitionId, MzOffset) {
+    let row = data.0.expect("invalid binding");
+    let mut datums = row.iter();
+    let (pid, offset_datum) = match (datums.next(), datums.next()) {
+        (Some(Datum::Int32(pid)), Some(offset_datum)) => (PartitionId::Kafka(pid), offset_datum),
+        (Some(offset_datum), None) => (PartitionId::None, offset_datum),
+        _ => panic!("invalid binding"),
+    };
+
+    let offset = match offset_datum {
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:77]: nit: these matches could be combined

File: src/storage/src/source/reclock.rs [hunk 2]
```diff
@@ -47,6 +47,91 @@ pub struct ReclockFollower {
...
+    (pid, offset)
+}
+
+/// Drains the provided vector of updates containing insertions and retractions of full offset
+/// statements and differentiates them into updates containing the change in offset in the diff
+/// field.
+fn differentiate(
+    updates: &mut Vec<((PartitionId, MzOffset), Timestamp, Diff)>,
+) -> impl Iterator<Item = (PartitionId, Timestamp, MzOffset)> + '_ {
+    consolidation::consolidate_updates(updates);
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:90]: is this strictly necessary? or does it just reduce the size of the remap trace

File: src/storage/src/source/reclock.rs [hunk 3]
```diff
@@ -631,9 +714,8 @@ impl ReclockOperator {
...
             let upper = self.upper.clone();
             let new_upper = new_upper.clone();
-            let updates = updates
-                .iter()
-                .map(|(pid, diff)| (((), pid.borrow()), next_ts, diff));
+            let updates = integrate(&self.source_upper, updates)
+                .map(|((pid, offset), diff)| ((pack_binding(pid, offset), ()), next_ts, diff));
             match self
                 .write_handle
                 .compare_and_append(updates, upper, new_upper)
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:721]: might want a comment here:

```
// We create a set of differential updates, all at the same timestamp,
// that _retract_ the previous source_upper (effectively zero-ing the collection),
// _insert_ the new upper
```

File: src/storage/src/source/reclock.rs [hunk 4]
```diff
@@ -47,6 +47,91 @@ pub struct ReclockFollower {
...
+                        assert_eq!(prev_pid, pid, "invalid bindings");
+                        assert_eq!(prev_ts, ts, "invalid bindings");
+                        prev_offset
+                    }
+                    None => MzOffset::from(0),
+                };
+                offset
+                    .checked_sub(prev_offset)
+                    .map(move |diff| (pid, ts, diff))
+            }
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:111]: might be nice to add a comment like "we effectively turn pairs of `source_upper` retractions+insertions (placed next to each other in the stream by `integrate` into mzoffset diffs"

**Review by guswynn — APPROVED** [review]
The idea of changing the storage format, but leaving the `remap_trace` and all that complexity in `reclock.rs` and `source_reader_pipeline.rs` alone is very good!

Are we comfortable forever locking into `SourceData` as the encoding (and using row's for backwards compat, not the niceties of protobuf), to make the remap-in-sql easier?

I also want to say that its nice this appears to not break the remap shard compaction i added, which is nice!

rest of my comments are nits, and mostly just comments that i can add later! merge away!

File: src/storage/src/source/reclock.rs (hunk 2 above)
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:90]: It is necessary for the updates to be sorted, but the consolidation will also reduce the work we need to do

File: src/storage/src/source/reclock.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:77]: indeed! combined

File: src/storage/src/source/reclock.rs (hunk 4 above)
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:111]: added

**Pull request merged**